src/sentry/data_secrecy/data_secrecy_logic.py

```diff
@@ -29,3 +34,121 @@ def should_allow_superuser_access(
+        return True
+
+    # If there is an active grant, allow superuser access
+    if data_access_grant_exists(organization.id):
+        return True
+
+    return False
+
+
+def _get_cached_grant_status(organization_id: int) -> bool | None:
+    """
+    Retrieve cached grant status for an organization.
+
+    :param organization_id: The ID of the organization to get the cached grant status for.
+    :return:
+        - False if negative cache hit.
+        - True if cached data is valid and not expired.
+        - None if not cached.
+    """
+    cache_key = CACHE_KEY_PATTERN.format(organization_id=organization_id)
+    cached_data = cache.get(cache_key)
+
+    if cached_data:
+        # Check if it's negative cache
+        if cached_data == NEGATIVE_CACHE_VALUE:
+            return False
+
+        # Verify the cached data hasn't logically expired
+        access_end = cached_data["access_end"]
+        if access_end > datetime.now(timezone.utc):
+            return True
+        else:
+            # Cached data is stale, remove it
+            cache.delete(cache_key)
+
+    return None
+
+
+def data_access_grant_exists(organization_id: int) -> bool:
+    """
+    Determines if a data access grant exists for an organization.
+    Tries cache first, calculates and caches if needed.
+
+    :param organization_id: The ID of the organization to check.
+    :return: True if a data access grant exists for the organization, False otherwise.
+    """
+    # Try cache first
+    cached_status = _get_cached_grant_status(organization_id)
+    if cached_status is not None:
+        return cached_status
+
+    # Cache miss or expired, calculate fresh
+    grant_status = data_access_grant_service.get_effective_grant_status(
+        organization_id=organization_id
+    )
+
+    if grant_status:
+        # Calculate TTL first to avoid race condition where grant expires between checks
+        current_time = datetime.now(timezone.utc)
+        ttl_seconds = int((grant_status.access_end - current_time).total_seconds())
+
+        # If the grant is expired or about to expire (TTL <= 0), cache the negative result
+        if ttl_seconds <= 0:
+            cache.set(
+                CACHE_KEY_PATTERN.format(organization_id=organization_id),
+                NEGATIVE_CACHE_VALUE,
+                timeout=NEGATIVE_CACHE_TTL,
+            )
+            return False
+
+        # Grant is still valid, cache the positive result with calculated TTL
+        serialized_grant_status = grant_status.dict()
+        cache.set(
+            CACHE_KEY_PATTERN.format(organization_id=organization_id),
+            serialized_grant_status,
+            timeout=ttl_seconds,
+        )
+        return True
+
+    # Cache the negative result for 15 minutes
+    else:
+        cache.set(
+            CACHE_KEY_PATTERN.format(organization_id=organization_id),
+            NEGATIVE_CACHE_VALUE,
+            timeout=NEGATIVE_CACHE_TTL,
```

Comment:
15 mins feels like a long time to prevent superuser access based on an initial miss. Could we shorten this maybe?